composites/Plugin/SnippetEditor/replaceText.js

```diff
@@ -0,0 +1,115 @@
+import { EditorState, SelectionState, Modifier } from "draft-js";
+
+/**
+ * Tests if a character in a text is whitespace. Out of range index counts as whitespace.
+ *
+ * @param {string} text  The text to get the character from.
+ * @param {number} index The index of the character in the text.
+ *
+ * @returns {boolean} True if the character is out of bounds or whitespace.
+ */
+export const hasWhitespaceAt = ( text, index ) => {
+	const char = text.charAt( index );
+
+	// No character found equals whitespace.
+	if ( char.length === 0 ) {
+		return true;
+	}
+
+	return /\s/.test( char );
+};
+
+/**
+ * Gets the Draft.js caret offset.
+ *
+ * @param {SelectionState} selection The Draft.js selection state.
+ *
+ * @returns {number} The Draft.js caret offset.
+ */
+export const getCaretOffset = ( selection ) => {
+	const isBackward = selection.getIsBackward();
+
+	// If the selection is backwards the end offset is the caret.
+	return isBackward ? selection.getEndOffset() : selection.getStartOffset();
+};
+
+/**
+ * Gets the block of the start of the current Draft.js selection.
+ *
+ * This is a tiny helper function to get the Draft.js selection anchor key and
+ * then the block of that key.
+ * Read more about the Draft.js selection state: https://draftjs.org/docs/api-reference-selection-state.html
+ * Read more about the Draft.js content block: https://draftjs.org/docs/api-reference-content-block.html
+ *
+ * @param {ContentState}   contentState   The Draft.js content state.
+ * @param {SelectionState} selectionState The Draft.js selection state.
+ *
+ * @returns {ContentBlock} The Draft.js ContentBlock.
+ */
+export const getAnchorBlock = ( contentState, selectionState ) => {
+	const key = selectionState.getAnchorKey();
+	return contentState.getBlockForKey( key );
+};
+
+/**
+ * Inserts text into the editor state.
+ *
+ * @param {EditorState} editorState The Draft.js editor state.
+ * @param {string}      text        The text to insert.
+ *
+ * @returns {EditorState} The new editor state.
+ */
+export const insertText = ( editorState, text ) => {
+	const contentState = editorState.getCurrentContent();
+	const selectionState = editorState.getSelection();
+
+	// If there is a selection instead of a caret do nothing.
+	if ( ! selectionState.isCollapsed() ) {
```

Comment:
Seems to me`selectionState.isCollapsed()` is always `true` whether there's a caret or a selection, so this condition never passes. Not sure why, also `document.getSelection()` tells me it's always `collapsed`. Maybe there's something going on before this condition that turns a selection to a caret?

Reply:
In `triggerReplacementVariableSuggestions` it removes the selected text before. So in the current use-case what you are saying is true. But when you use the `insertText` without that check it would error if it is not collapsed.